soteria-rust/lib/layout.ml

```diff
@@ -189,63 +196,87 @@ let rec layout_of (ty : Types.ty) : layout =
       | Struct fields -> layout_of_struct adt fields
       | Enum variants -> layout_of_enum adt variants
       | Union fs ->
-          let layouts = List.map layout_of @@ Charon_util.field_tys fs in
+          let++ layouts =
+            Result.fold_list (Charon_util.field_tys fs) ~init:[]
+              ~f:(fun acc ty ->
+                let++ l = layout_of ty in
+                l :: acc)
+          in
+          let layouts = List.rev layouts in
+
           let hd = List.hd layouts in
           let tl = List.tl layouts in
           let size, align =
             List.fold_left
-              (fun (size, align) l -> (max size l.size, max align l.align))
+              (fun (size, align) l ->
+                ( BV.max ~signed:false size l.size,
+                  BV.max ~signed:false align l.align ))
               (hd.size, hd.align) tl
           in
           let size = size_to_fit ~size ~align in
           (* All fields in the union start at 0 and overlap *)
-          let fields = Array.make (List.length fs) 0 in
-          { size; align; fields = Arbitrary (Types.VariantId.zero, fields) }
-      | Opaque ->
-          let msg = Fmt.str "opaque (%a)" Crate.pp_name adt.item_meta.name in
-          raise (CantComputeLayout (msg, ty))
-      | TDeclError _ -> raise (CantComputeLayout ("decl error", ty))
-      | Alias _ -> raise (CantComputeLayout ("alias", ty)))
+          let fields = Array.make (List.length fs) (BV.usizei 0) in
+          mk ~size ~align ~fields:(Arbitrary (Types.VariantId.zero, fields)) ()
+      | Opaque -> not_impl_layout "opaque" ty
+      | TDeclError _ -> not_impl_layout "decl error" ty
+      | Alias _ -> not_impl_layout "alias" ty)
   (* Arrays *)
   | TAdt { id = TBuiltin TArray; generics } ->
+      let max_array_len sub_size =
+        (* We calculate the max array size for a 32bit architecture, like Miri does. *)
+        if Typed.equal sub_size Usize.(0s) then
+          (* Special case to avoid a division by zero in the concrete case *)
+          BV.usize Z.(pow (of_int 2) 32)
+        else
+          let isize_bits = 32 - 1 in
+          BV.usize Z.(one lsl isize_bits) /@ Typed.cast sub_size
```

Comment:
How is this sound?

Reply:
> note that this isn't a problem when sub_layout.size is symbolic and evaluates to 0 since it makes the condition true anyways, regardless of the / 0.

That was my main issue yeah. As you say, it's prob possible to re-organise code so that it is clear that the div-by-zero case is not unsound because it's not used